### PR TITLE
Support Global Doc V

### PR DESCRIPTION
**Pull request by jumaallan**
Story: https://app.shortcut.com/smileid/story/10425/support-global-docv

## Summary

A few sentences/bullet points about the changes

## Known Issues

Any shortcomings in your work. This may include corner cases not correctly handled or issues related
to but not within the scope of your PR. Design compromises should be discussed here if they were not
already discussed above.

## Test Instructions

Concise test instructions on how to verify that your feature works as intended. This should include
changes to the development environment (if applicable) and all commands needed to run your work.

## Screenshot

If applicable (e.g. UI changes), add screenshots to help explain your work.

### CODE REVIEW

File: lib/src/main/java/com/smileidentity/compose/SmileIDExt.kt [hunk 1]
```diff
@@ -146,7 +145,8 @@ fun SmileID.SmartSelfieAuthentication(
  */
 @Composable
 fun SmileID.DocumentVerification(
-    idType: Document,
+    countryCode: String? = null,
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/compose/SmileIDExt.kt:148]: Shouldn't the countryCode be required?

Can you also update the KDocs above?

Comment (jumaallan) [reply lib/src/main/java/com/smileidentity/compose/SmileIDExt.kt:148]: yeah let me fix this

**Comment by vanshg** [comment]
Can you also update the changelog?

File: lib/src/main/java/com/smileidentity/fragment/DocumentVerificationFragment.kt [hunk 2]
```diff
@@ -71,7 +70,8 @@ class DocumentVerificationFragment : Fragment() {
             jobId: String = randomJobId(),
             showAttribution: Boolean = true,
             allowGalleryUpload: Boolean = false,
-            idType: Document,
+            countryCode: String,
+            documentType: String,
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/fragment/DocumentVerificationFragment.kt:74]: Can you make these 2 the first 2 parameters of the method? Since they are required arguments

File: lib/src/main/java/com/smileidentity/fragment/DocumentVerificationFragment.kt [hunk 3]
```diff
@@ -140,10 +142,15 @@ private var Bundle.allowGalleryUpload: Boolean
     get() = getBoolean(KEY_ALLOW_GALLERY_UPLOAD)
     set(value) = putBoolean(KEY_ALLOW_GALLERY_UPLOAD, value)
 
-private const val KEY_ID_TYPE = "idType"
-private var Bundle.idType: Document
-    get() = getParcelableCompat(KEY_ID_TYPE)!!
-    set(value) = putParcelable(KEY_ID_TYPE, value)
+private const val KEY_COUNTRY_CODE = "countryCode"
+private const val KEY_DOCUMENT_TYPE = "documentType"
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/fragment/DocumentVerificationFragment.kt:146]: Can you move this line down to be with the declaration of the documentType extension property?

File: lib/src/main/java/com/smileidentity/viewmodel/document/OrchestratedDocumentViewModel.kt [hunk 4]
```diff
@@ -133,7 +133,8 @@ internal class OrchestratedDocumentViewModel(
                     backImageInfo,
                     selfieImageInfo,
                 ) + livenessImageInfo,
-                idInfo = IdInfo(idType.countryCode, idType.documentType),
+                idInfo = IdInfo(country = countryCode, idType = documentType ?: ""),
+                // todo - what happens when we don't pass a document type here?
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/viewmodel/document/OrchestratedDocumentViewModel.kt:137]: Shouldn't IdInfo be updated to allow for a null idType? or does it expect an empty string?

File: sample/src/main/java/com/smileidentity/sample/compose/DocumentVerificationIdTypeSelector.kt [hunk 5]
```diff
@@ -205,7 +203,13 @@ fun DocumentVerificationIdTypeSelector(
             }
         }
         Button(
-            onClick = { onIdTypeSelected(selectedCountry!!, selectedIdType!!) },
+            // onClick = { onIdTypeSelected(selectedCountry!!, selectedIdType) },
```
Comment (vanshg) [inline sample/src/main/java/com/smileidentity/sample/compose/DocumentVerificationIdTypeSelector.kt:206]: Remove commented code

File: sample/src/main/java/com/smileidentity/sample/compose/MainScreen.kt [hunk 6]
```diff
@@ -235,31 +234,28 @@ fun MainScreen(
...
                     SmileID.DocumentVerification(
                         userId = userId,
                         jobId = jobId,
-                        idType = documentType,
+                        countryCode = it.arguments?.getString("countryCode")!!,
+                        documentType = it.arguments?.getString("documentType"),
                         showInstructions = true,
                         allowGalleryUpload = true,
-                        captureBothSides = true,
+                        captureBothSides = it.arguments?.getBoolean("hasBackSide")!!,
```
Comment (vanshg) [inline sample/src/main/java/com/smileidentity/sample/compose/MainScreen.kt:258]: I think it may be better to remove this from the interface entirely, now that we have the API endpoint.

Also, it should be the DocV flow that automatically determines whether there is a backside, it shouldn't need to be passed in.

And in the case where only the country code is supplied and we can't check whether this is a back or not, we need to show a skip button on the Document Back Instructions screen

**Review by vanshg — CHANGES_REQUESTED** [review]
The call to `hasBackSide` should happen _within_ the DocV flow (so that it is abstracted away from partners) rather than as a parameter that is passed in

File: lib/src/main/java/com/smileidentity/compose/SmileIDExt.kt [hunk 7]
```diff
@@ -147,9 +147,9 @@ fun SmileID.SmartSelfieAuthentication(
...
 @Composable
 fun SmileID.DocumentVerification(
-    idType: Document,
+    countryCode: String,
+    documentType: String? = null,
     idAspectRatio: Float? = null,
-    captureBothSides: Boolean = false,
     bypassSelfieCaptureWithFile: File? = null,
     userId: String = rememberSaveable { randomUserId() },
     jobId: String = rememberSaveable { randomJobId() },
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/compose/SmileIDExt.kt:155]: while you're here can you change `showInstructions` to default to `true`? (not sure why it was set to false initially)

File: lib/src/main/java/com/smileidentity/networking/SmileIDService.kt [hunk 8]
```diff
@@ -113,6 +114,12 @@ interface SmileIDService {
     @POST("/v1/products_config")
     suspend fun getProductsConfig(@Body request: ProductsConfigRequest): ProductsConfigResponse
 
+    /**
+     * Returns Global DocV supported products and metadata
+     */
+    @POST("/v1/valid_documents")
+    suspend fun getGlobalDocVConfig(@Body request: ProductsConfigRequest): GlobalDocVResponse
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/networking/SmileIDService.kt:121]: nit: getValidDocuments

File: lib/src/main/java/com/smileidentity/viewmodel/document/OrchestratedDocumentViewModel.kt (hunk 4 above)
Comment (jumaallan) [reply lib/src/main/java/com/smileidentity/viewmodel/document/OrchestratedDocumentViewModel.kt]: This is fixed

File: lib/src/main/java/com/smileidentity/fragment/DocumentVerificationFragment.kt (hunk 3 above)
Comment (jumaallan) [reply lib/src/main/java/com/smileidentity/fragment/DocumentVerificationFragment.kt]: Done

**Comment by jumaallan** [comment]
> The call to `hasBackSide` should happen _within_ the DocV flow (so that it is abstracted away from partners) rather than as a parameter that is passed in

As per the last discussion, the call now happens inside the country selection flow, and is passed into the sdk as a parameter. The value is always set to true, so we can show the skip button inside the sdk

File: sample/src/main/java/com/smileidentity/sample/viewmodel/IdTypeSelectorAndFieldInputViewModel.kt [hunk 9]
```diff
@@ -89,19 +88,19 @@ class IdTypeSelectorAndFieldInputViewModel(
...
-                    countryDetails[it.countryCode] ?: SearchableInputFieldItem(
-                        it.countryCode,
-                        it.name,
-                    )
-                }
-                .sortedBy { it.displayName }
-                .toImmutableList()
-
-            _uiState.update { it.copy(countries = countryList) }
+            // val countryList = servicesResponseForJobType
```
Comment (vanshg) [inline sample/src/main/java/com/smileidentity/sample/viewmodel/IdTypeSelectorAndFieldInputViewModel.kt:91]: Let's clean up the commented code

File: sample/src/main/java/com/smileidentity/sample/compose/DocumentVerificationIdTypeSelector.kt [hunk 10]
```diff
@@ -61,155 +55,117 @@ import timber.log.Timber
 fun DocumentVerificationIdTypeSelector(
     modifier: Modifier = Modifier,
     viewModel: DocumentSelectorViewModel = viewModel(),
-    onIdTypeSelected: (String, String?) -> Unit,
+    onIdTypeSelected: (String, String?, Boolean) -> Unit,
```
Comment (vanshg) [inline sample/src/main/java/com/smileidentity/sample/compose/DocumentVerificationIdTypeSelector.kt:58]: can we give names to these?
`onIdTypeSelected: (countryCode: String, idType: String?, captureBothSides: Boolean) -> Unit,`

File: CHANGELOG.md [hunk 11]
```diff
@@ -22,6 +22,7 @@
 - Detection of bad lighting for Document Verification
 - Detection of unfocused states for Document Verification
 - Document detection for Document Verification
+- DocV updated to dynamically shows back of ID capture
```
Comment (vanshg) [inline CHANGELOG.md:25]:
```suggestion
- Document Verification updated to dynamically shows back of ID capture 
```

File: CHANGELOG.md [hunk 12]
```diff
@@ -38,6 +39,9 @@
 - Bump AndroidX Navigation to 2.7.1
 
 ### Removed
+- Removed `documentType` as a required field in DocV
```
Comment (vanshg) [inline CHANGELOG.md:42]:
```suggestion
- Removed `documentType` as a required field in Document Verification
```

File: CHANGELOG.md [hunk 13]
```diff
@@ -38,6 +39,9 @@
...
+- Removed `Document` model, so you now pass `countryCode` and `documentType` as separate params in
+  DocV
```
Comment (vanshg) [inline CHANGELOG.md:44]:
```suggestion
  Document Verification
```

File: lib/src/main/java/com/smileidentity/models/ProductsConfig.kt [hunk 14]
```diff
@@ -30,3 +30,39 @@ data class IdSelection(
     @Json(name = "enhanced_kyc") val enhancedKyc: IdTypes = emptyMap(),
     @Json(name = "doc_verification") val documentVerification: IdTypes = emptyMap(),
 )
+
+@JsonClass(generateAdapter = true)
+data class GlobalDocVResponse(
```
Comment (vanshg) [inline lib/src/main/java/com/smileidentity/models/ProductsConfig.kt:35]: Can you rename this to `ValidDocumentsResponse`

File: lib/src/main/java/com/smileidentity/networking/SmileIDService.kt (hunk 8 above)
Comment (vanshg) [reply lib/src/main/java/com/smileidentity/networking/SmileIDService.kt:121]: bump

File: lib/src/main/java/com/smileidentity/compose/document/OrchestratedDocumentVerificationScreen.kt [hunk 15]
```diff
@@ -57,7 +57,7 @@ internal fun OrchestratedDocumentVerificationScreen(
             showInstructions = showInstructions,
             showAttribution = showAttribution,
             allowGallerySelection = allowGalleryUpload,
-            captureBothSides = captureBothSides,
+            showSkipButton = false,
```
Comment (jumaallan) [inline lib/src/main/java/com/smileidentity/compose/document/OrchestratedDocumentVerificationScreen.kt:60]: Do we want to set it to `false` here explicitly? 

Or should we use the `captureBothSides` flag?

Comment (vanshg) [reply lib/src/main/java/com/smileidentity/compose/document/OrchestratedDocumentVerificationScreen.kt:60]: It should be false, since this is for the front of document

Comment (jumaallan) [reply lib/src/main/java/com/smileidentity/compose/document/OrchestratedDocumentVerificationScreen.kt:60]: And the back side? Thats set to true by default

The comment was for the two sides actually (Front and Back)

Comment (vanshg) [reply lib/src/main/java/com/smileidentity/compose/document/OrchestratedDocumentVerificationScreen.kt:60]: I don't think so, I see `showSkipButton = captureBothSides,` for the back side

Comment (jumaallan) [reply lib/src/main/java/com/smileidentity/compose/document/OrchestratedDocumentVerificationScreen.kt:60]: I actually didn't see that 😆

**Review by vanshg — APPROVED** [review]
